File: app/src/main/java/net/tlalka/android/fiszki/activities/LessonActivity.java
package net.tlalka.android.fiszki.activities;

import android.os.Bundle;
import android.view.View;
import android.widget.Button;
import android.widget.TextView;
import butterknife.BindView;
import butterknife.OnClick;
import net.tlalka.android.fiszki.R;
import net.tlalka.android.fiszki.fragments.LanguageDialogFragment;
import net.tlalka.android.fiszki.models.dto.LessonDto;
import net.tlalka.android.fiszki.models.entities.Word;
import net.tlalka.android.fiszki.models.types.LanguageType;
import net.tlalka.android.fiszki.models.types.StorageType;
import net.tlalka.android.fiszki.navigations.Navigator;
import net.tlalka.android.fiszki.services.LessonService;
import net.tlalka.android.fiszki.services.StorageHelper;
import net.tlalka.android.fiszki.utils.ValidUtils;

import javax.inject.Inject;
import java.util.List;

public class LessonActivity extends BasePageActivity implements LanguageDialogFragment.DialogListener {

    @BindView(R.id.text_view_topic)
    protected TextView textViewTopic;

    @BindView(R.id.show_word_button)
    protected Button buttonWordShow;

    @BindView(R.id.check_word_button)
    protected Button buttonWordCheck;

    @Inject
    protected LessonService lessonService;

    @Inject
    protected StorageHelper storageHelper;

    @Inject
    protected Navigator navigator;

    @Inject
    protected LessonDto lessonDto;

    private LanguageType translation;

    @Override
    public void onCreate(Bundle bundle) {
        super.onCreate(bundle);
        super.setContentView(R.layout.lesson_activity);
        super.getActivityComponent().inject(this);

        this.initStorage();
        this.runActivity();
    }

    private void initStorage() {
        this.translation = this.storageHelper.getEnum(StorageType.TRANSLATION, LanguageType.PL);
    }

    private void runActivity() {
        String lessonName = this.lessonDto.getLessonName();
        String lessonDesc = this.lessonDto.getLessonLevel().name().toLowerCase();

        this.textViewTopic.setText(localFormat("%s - %s", lessonName, lessonDesc));
        this.generateView();
    }

    private void generateView() {
        if (this.lessonService.hasNextWord()) {
            this.buttonWordShow.setText(this.lessonService.getNextWord());
            this.buttonWordCheck.setText(getText(R.string.lesson_activity_check_word));
        } else {
            this.showLessonSummary();
        }
    }

    private void showLessonSummary() {
        this.lessonService.updateLessonProgress();
        this.navigator.openLessonStatActivity(this, this.lessonDto);
        this.navigator.finish(this);
    }

    @OnClick(R.id.check_word_button)
    public void onCheckWordClick(View view) {
        Word word = this.lessonService.getTranslation(this.translation);

        if (ValidUtils.isNotNull(word)) {
            this.buttonWordCheck.setText(word.getValue());
        } else {
            List<LanguageType> languages = this.lessonService.getLanguages();
            languages.remove(this.translation);

            LanguageDialogFragment dialog = LanguageDialogFragment.getInstance(languages);
            dialog.show(getFragmentManager(), LanguageDialogFragment.class.getName());
        }
    }

    @Override
    public void onLanguageSelected(LanguageType languageType) {
        this.translation = languageType;
        this.onCheckWordClick(this.buttonWordCheck);
    }

    @OnClick(R.id.button_good)
    public void onCorrectClick(View view) {
        this.lessonService.correctAnswer();
        this.generateView();
    }

    @OnClick(R.id.button_bad)
    public void onIncorrectClick(View view) {
        this.lessonService.incorrectAnswer();
        this.generateView();
    }
}

File: app/src/main/kotlin/net/tlalka/fiszki/view/activities/LessonActivity.kt
package net.tlalka.fiszki.view.activities

import android.os.Bundle
import android.view.View
import android.widget.Button
import android.widget.TextView
import butterknife.BindView
import butterknife.OnClick
import net.tlalka.fiszki.R
import net.tlalka.fiszki.domain.controllers.LessonController
import net.tlalka.fiszki.domain.utils.ValidUtils
import net.tlalka.fiszki.model.dto.parcel.LessonDto
import net.tlalka.fiszki.model.types.LanguageType
import net.tlalka.fiszki.view.fragments.LanguageDialogFragment
import net.tlalka.fiszki.view.navigations.Navigator
import javax.inject.Inject

class LessonActivity : BasePageActivity(), LanguageDialogFragment.DialogListener {

    @BindView(R.id.lesson_topic)
    lateinit var lessonTopic: TextView

    @BindView(R.id.lesson_progress)
    lateinit var lessonProgress: TextView

    @BindView(R.id.lesson_show_word)
    lateinit var lessonShowWord: Button

    @BindView(R.id.lesson_check_word)
    lateinit var lessonCheckWord: Button

    @Inject
    lateinit var lessonController: LessonController

    @Inject
    lateinit var navigator: Navigator

    @Inject
    lateinit var lessonDto: LessonDto

    public override fun onCreate(bundle: Bundle?) {
        super.onCreate(bundle)
        super.setContentView(R.layout.lesson_activity)
        super.activityComponent.inject(this)

        runActivity()
    }

    private fun runActivity() {
        lessonTopic.text = getString(R.string.lesson_topic, lessonDto.lessonIndex, lessonDto.lessonName)
        generateView()
    }

    private fun generateView() {
        if (lessonController.hasNextWord()) {
            lessonProgress.text = lessonController.getLessonStatus()
            lessonShowWord.text = lessonController.getNextWord()
            lessonCheckWord.text = getText(R.string.lesson_check_word)
        } else {
            showLessonSummary()
        }
    }

    private fun showLessonSummary() {
        lessonController.updateLessonDto(lessonDto)
        lessonController.updateLessonProgress()
        navigator.openLessonScoreActivity(this, lessonDto)
        navigator.finish(this)
    }

    @OnClick(R.id.lesson_check_word)
    fun onCheckWordClick(@Suppress("UNUSED_PARAMETER") view: View?) {
        val word = this.lessonController.getTranslateWord()

        if (ValidUtils.isNotNull(word)) {
            lessonCheckWord.text = word.value
        } else {
            LanguageDialogFragment
                    .getInstance(lessonController.getLanguages())
                    .show(fragmentManager, LanguageDialogFragment::class.java.name)
        }
    }

    override fun onLanguageSelected(languageType: LanguageType) {
        lessonController.setTranslation(languageType)
        onCheckWordClick(lessonCheckWord)
    }

    @OnClick(R.id.button_correct)
    fun onCorrectClick(@Suppress("UNUSED_PARAMETER") view: View) {
        lessonController.correctAnswer()
        generateView()
    }

    @OnClick(R.id.button_incorrect)
    fun onIncorrectClick(@Suppress("UNUSED_PARAMETER") view: View) {
        lessonController.incorrectAnswer()
        generateView()
    }
}
